Add overlay2.size daemon storage-opt

imkin opened the pull request:
This commit adds the overlay2.size option to the daemon daemon storage opts.

The user can override this option by the "docker run --storage-opts" options.

Signed-off-by: Dhawal Yogesh Bhanushali <dbhanushali@vmware.com>

closes https://github.com/moby/moby/issues/30491

**- What I did**

Added the daemon storage opt overlay2.size option

docker daemon --storage-opt overlay2.size=5G

imkin commented:
@GordonTheTurtle @amir73il any comments?

AkihiroSuda commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -191,9 +191,12 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 		}
 	}
-
+	//
```
extra change

imkin replied:
Made the change.

AkihiroSuda commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -191,9 +191,12 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 			projectQuotaSupported = true
+		} else if opts.quota.Size > 0 {
+			logrus.Error("projectQuota Not supported")
+			return nil, err
```
please wrap err with a human-friendly text

imkin replied:
Do you suggest to have something like graphdriver.ErrPrjQuotaNotSupported ? or graphdriver.FeatureNotSupported ? 

Please suggest a name otherwise. Thanks :-)

imkin replied:
Looked at some of the other drivers and made the change. Thanks

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -343,6 +351,11 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
+			// Set default container disk quota limit
+			if err := d.quotaCtl.SetQuota(dir, d.options.quota); err != nil {
+				return err
+			}
```
One problem with the way this code is structured is that it doesn't allow the user to override the default size with 0 (i.e. no quota for container).

Another problem is that it is just unneeded code duplication.
How about setting 
driver.options.quota = d.options.quota
just before d.parseStorageOpt

This will also be generic enough to apply future defaults if Quota struct ever
gets values like soft limit, inode limit etc...

imkin replied:
Thanks for the suggestion. Made the change.

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -213,7 +216,12 @@ func parseOptions(options []string) (*overlayOptions, error) {
 				return nil, err
 			}
-
+		case "overlay2.default_size":
```
Because the storage option name in create/start command is the same for all graphdrivers that support it (i.e. "size=") I think the option name for dockerd storage option should be equally generic (i.e. "default_size="). This will also make documentation more generic. BTW, how about the documentation?

imkin replied:
Added the documentation portion.

Made the change to the option being called as overlay2.size instead of overlay2.default_size

amir73il replied:
So you removed the "default_" part and not the "overlay2." part. Oh well, I guess there are arguments to justify both alternatives. Either are fine by me.

AkihiroSuda commented on docs/reference/commandline/dockerd.md:
````diff
@@ -685,6 +685,19 @@ to add multiple lower directory support for OverlayFS. This option should
+###### Example
+
+```bash
+$ sudo dockerd -s overlayfs2 --storage-opt overlay2.size=10G
````
overlayfs2 -> overlay2

imkin replied:
ah thanks fixed.

amir73il reviewed (approved):
LGTM

rhvgoyal commented:
Does this apply to only containers or to image as well?

rhvgoyal commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -213,7 +214,12 @@ func parseOptions(options []string) (*overlayOptions, error) {
+			if err != nil {
+				return nil, err
+			}
+			o.quota.Size = uint64(size)
```
I think we should if  we are on xfs and we can enable quota. Otherwise we should fail during Init() time itself if user wants to use a default size but we can't enforce that default size.

rhvgoyal replied:
IOW, in Init() this check should be there and driver initialization should fail if quota can't be supported and user asked for it.

amir73il replied:
Right. I missed that.

imkin replied:
In sync with the other comments have made the change.

bklau replied:
@amir73il Q: so the overlay2.size storage opt sets the default for RW layer only or includes the base image(Rootfs)?
Pls clarify.

amir73il replied:
Only for RW layer.

amir73il commented:
The quota is only applied to container root dir, which is where the overlay upper dir is,
so unlike btrfs and devicemapper drivers, 'size' doesn't limit the size of the image at all,
only the size of the changes that container can make to the image.
However that is the 'size' that will be observed by user running 'df /' command inside container.

rhvgoyal commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -191,11 +192,11 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
 		if d.quotaCtl, err = quota.NewControl(home); err == nil {
 			projectQuotaSupported = true
+		} else if opts.quota.Size > 0 {
+			return nil, fmt.Errorf("Project quota is not supported: %v", err)
```
Oh, I see, here is the check and error out if quota is not supported. But this will run only if we are running on to of xfs.

How about moving this check out of backingFs == "xfs" if block.

Also error message could probably be more verbose. Say.

If we are not running on xfs, that could be first check and error message could be.

`Storage option overlay2.size not supported. Underlying filesystem needs to be xfs`

And if we are running on xfs but not supporting quota, we could say.

`Storage option overlay2.size not supported. Filesystem is does not support project quota`

amir73il replied:
As @rhvgoyal noted correctly, this test should move outside the backingFs == "xfs" block
(i.e. if  !projectQuotaSupported) and error message should be made more clear to the user, something like the error in Create():
"overlay2.size option is supported only for overlay over xfs with 'pquota' mount option"

amir73il replied:
That was a race in review :) I did not notice you posted this. Agreed.
Typo "Filesystem is does not ..."

imkin replied:
Made the change as suggested. The messages are little different but mean the same.

imkin replied:
Added as per @rhvgoyal suggestions.

rhvgoyal commented:
@amir73il We seem to enforce quota restrictions in Create(). And IIUC, that's used both by images as well as container root. So IIUC, this quota will be applied to both images as well as container (after this patch).

Previously, it will not apply to images because option was per container and there was no way to pass this option for images.

Am I missing something?

rhvgoyal commented:
I think applying these limits for images is a good idea though. People have raised concerns about pulling images with sparse files in it. And being able to put a limit on the max size of each layer will help prevent somebody trying to explode 1TB sparse file on disk.

amir73il commented:
I see. confession: I don't know docker that well... I guess applying limits to images make sense. So an image will contain single or multi lower layers, which will all limited cumulatively by the project quota limit?
Specifically, I also don't know what the container '-init' dir is, but it gets assigned a project id as well, so I am guessing it will get a project default quota limit too now.

rhvgoyal commented:
Right. -init is another intermediate layer where docker might want to scribble something and then use it as lower layer for actual container rootfs. So quota will be applied to that too. Though contents written there by docker are very small, so applying quota to -init layer is not a problem.

I think being able to apply it to container rootfs and to image layer are going to be very useful.

rhvgoyal commented:
That's my understanding that every image layer will be limited to project quota specified by overlay2.size. But I also have not spent too much time in looking at overlay2 code. So @imkin might have more definitive answer.

thaJeztah commented:
/cc @dmcgowan

imkin commented:
@rhvgoyal I do see that all the layers comply with the "size" parameter. 

This, however, is same as what the existing size storage option for the docker run command. So I believe if anything else needs to comply with this "size" param then it should be a holistic issue of the overlayfs2's docker driver here and not in the scope of this PR.

imkin commented:
The powerpc test failure is a flake.

imkin commented:
@amir73il Made the suggested changes. Please bless :-)

rhvgoyal commented:
@imkin, I am just trying to understand the behavior better. Previously though it was in create path, there was no way to pass in this during create for images so it did not impact us.

But with your changes, this being a daemon flag, it will apply to both image layer create and container rootfs. And it will impact image pull. 

Given your changes impact that behavior I want to understand the impact on images. 

So this will pull size restriction on image layer as well, right? Only thing seems to be that size limit will
be per image layer  (and not the whole image if image has multiple layers).

amir73il reviewed (approved):
code looks fine to me. still need to understand the implication of image pull. and there is a typo in commit message (daemon daemon)

imkin commented:
@rhvgoyal just now did a docker pull of a image larger than what the overlay2.size parameter allowed.

bash-4.3# docker pull openjdk
Using default tag: latest
latest: Pulling from library/openjdk
cd0a524342ef: Extracting [==================================================>] 52.55 MB/52.55 MB
e39c3ffe4133: Download complete 
85334a7c2001: Download complete 
b46c5b79125e: Download complete 
30d4cb7cc8bc: Download complete 
4d273117faaf: Download complete 
2b271bed4ca3: Download complete 
30e4d8ac7a54: Download complete 
45858d68b18d: Download complete 
failed to register layer: Error processing tar file(exit status 1): open /usr/share/locale/nb/LC_MESSAGES/tar.mo: disk quota exceeded

So @rhvgoyal @amir73il  it docker pull also honors the size parameter. The pull is successful when I increase the value of the "size" param.

rhvgoyal commented:
@imkin, thanks for verifying. I am glad that it works for image layer too. 

LGTM

runcom commented:
Awesome

runcom commented:
I guess design looks to to me as well, @thaJeztah @cpuguy83 can we move this to code review?

dmcgowan commented:
This quota should not apply to a pulled layer. Any restrictions related to pulling an image can and should be done before putting data in the graph driver. Adding this restrictions would be out of scope for what this change is trying to accomplish.

rhvgoyal commented:
@dmcgowan Is there a reason that size restriction should not apply to image layer. I mean if we get it by default, then trying to get away with that might not be helpful.

So once concern people have raised is what happens if I manage to generate images with sparse files and when I try to explode that image, that might fill up all my storage.  AFAIK, we have not figured out yet how to handle sparse files.

If that's the case, then putting this restriction on image layer as well, will be very helpful.

So all I am saying is that it is common code which works both for image layer and container rootfs. And trying to dilute it to skip applying quota only on image layer might reduce the usefulness of this knob.

Until and unless you are thinking of a separate knob for image layer restrictions. 

IMHO, a single knob controlling both container size and image size is not probably a bad idea. I think only if here is that it applies to the layer size and not whole image size. So one can still push lot of layers and overall consume quite a some amount of space on disk.

dmcgowan commented:
>  Is there a reason that size restriction should not apply to image layer. 

It is overloading this option. Currently the size option can only be provided on run and only applies to the container RW layer. We want to avoid this kind of diverging behavior based on the graph driver. If we want to implement limitations on images being pulled, that could be done through accounting on tar extraction and apply to every graph driver. Even with sparse files, the file size has to be declared in the tar header.

> Until and unless you are thinking of a separate knob for image layer restrictions.

I think it is a separate knob and a completely different feature. The container quota is intended to limit the space given to an application running inside the container, a layer restriction is about managing images on the host.

amir73il commented:
Is there a reason that size restriction should not apply to image layer.

It is overloading this option. Currently the size option can only be
provided on run and only applies to the container RW layer. We want to
avoid this kind of diverging behavior based on the graph driver.


Isn't that behavior the same as what you get from devicemapper.basesize
daemon storage option? Doesn't the basesize apply to images?
Sorry, but I have no idea about devicemapper graphdriver internals, but
documentation suggests that basesize limits image size.

imkin commented:
both the failures of the tests are the same flakes as reported above. https://github.com/moby/moby/issues/33041

rhvgoyal commented:
@amir73il Right, basesize in devicemapper applies to images as well. It determines the size of initial block device and rest of the block devices are snapshots of that device (both image and containers). So it ends up limiting the size of both image and container rw layer.

thaJeztah commented:
> Right, basesize in devicemapper applies to images as well.

IIRC, that was a limitation of the way it worked for devicemapper, not an intended part of the feature, but if that's not the case, I misunderstood that feature.

rhvgoyal commented:
I think for devicemapper we don't have a choise. basesize decides the initial size of device and that applies to both images and containers. If we were to not enforce that limit for images, then we will have to work extra hard to grow block device and filesystem whenever filesystem is about to get full.

I feel it will be better if this size knob applies both for ROLayer and RWLayer. It will be more consistent with devicemapper. And nowhere in the naming of field we imply that it applies only to RWLayer only.

Making this knob fine grained will only help if there is a use case where we want to allow pulling in bigger images but allow a smaller quota for RWLayer. Is that a common case. If that turns out to be a problem, then we can introduce another knob which applies to RWLayer only. Say `overlay2.rwlayer_size`

cpuguy83 commented:
I don't think it makes any sense to limit the size of RO layers since it effectively does nothing to limit the overall size of the image.

imkin commented:
If we want to do this for the readwrite layer alone. does any one has any suggestions on how to identify this in the Create method?

cpuguy83 commented:
There's already existing code to set this on the RW layer only.
But the place to start would really be `CreateReadWrite` rather than
`Create`

On Tue, May 9, 2017 at 8:11 PM, Dhawal Yogesh Bhanushali <
notifications@github.com> wrote:

> If we want to do this for the readwrite layer alone. does any one has any
> suggestions on how to identify this in the Create method?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/moby/moby/pull/32977#issuecomment-300336831>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAwxZkI-U7E18Ct7a3YyEGft72NxXH8Kks5r4QDGgaJpZM4NOx3q>
> .
>



-- 


- Brian Goff

Commit:
Add overlay2.size daemon storage-opt

This commit adds the overlay2.size option to the daemon daemon
storage opts.

The user can override this option by the "docker run --storage-opt"
options.

Signed-off-by: Dhawal Yogesh Bhanushali <dbhanushali@vmware.com>

imkin commented:
@AkihiroSuda @rhvgoyal made the changes as per the latest discussions.

imkin commented:
@cpuguy83 @stevvooe Made the changes as per the latest discussions.

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -312,23 +322,33 @@ func (d *Driver) Cleanup() error {
-	if opts != nil && len(opts.StorageOpt) != 0 && !projectQuotaSupported {
-		return fmt.Errorf("--storage-opt is supported only for overlay over xfs with 'pquota' mount option")
-	}
-
```
I would leave this check in this function, because this is the function that does SetQuota. by moving the test to CreateReadWrite you put an implicit assumption that Create cannot be called directly with storage options, which is only true with current code.

imkin replied:
This is projectQuota support check and we want the project quota support only for the readwrite layer. Hence the check was moved.

amir73il replied:
IF you would have moved the code of setting pquota to CreateReadWrite() moving the test would have made sense, but the pquota code stayed in this function (for a good reason) so the "assert" test should stay here as well and then we don't need to ASSUME that Create() cannot be called with storage options except from CreateReadWrite().

amir73il replied:
Right. got it.

thaJeztah replied:
So if we don't want `Create` to be called with quota; perhaps we should add a non-exported `create()` that's both called by `Create()` and `CreateReadWrite()`;

```go
func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr error) {
	if opts != nil && len(opts.StorageOpt) != 0 && !projectQuotaSupported {
		return fmt.Errorf("--storage-opt is supported only for overlay over xfs with 'pquota' mount option")
	}

	return d.create(id, parent string, opts)
}
```

The check in `Create()` could even be;

```go
if opts != nil && len(opts.StorageOpt) != 0 {
    // error...
}
```

Given that `Create()` then won't support any options

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -343,20 +363,23 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
-			}
 		}
 	}
-
```
no reason to remove there newlines..

imkin replied:
Ah Yes :-)

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -312,23 +322,33 @@ func (d *Driver) Cleanup() error {
+
+	if len(opts.StorageOpt) == 0 {
+		opts.StorageOpt = map[string]string{
+			"size": fmt.Sprintf("%d", d.options.quota.Size),
```
Yes, size is currently the only overlay storage option, but options.quota can potentially gain more fine grained parameters , like hard/soft limit of blocks/inodes.
The code in Create() that overrides the driver defaults with storage options passed into run/create command already takes that into account.
This code should not assume that "size" is the only storage option. And why is this code needed at all? maybe I am missing something

imkin replied:
This assumption of "size" param was in the old code. The parsing of the options is done in Create. So In CreateReadWrite we do not know what the "options" are (we would need to parse and unparse) and in Create we do not know if it is for readWrite only. 

Create is called for read layers and readwrite layers. The current discussions evolved to having "the size" option only for readwrite layer. Hence the changes.

amir73il replied:
I understand the motivation, I just don't understand how this code helps or even does the right thing to prevent default quota from being applied on ReadOnly layers.
in Create(), the default driver.options.quota.Size is checked regardless of whether you manipulate StorageOpt here or not. Am I wrong?

amir73il replied:
ok I see what's going on here. this code makes the default handling code in Create() redundant.
So at the very least, redundant code should be removed.
In fact, I think that with this code you should loose *all* your changes to Create() function.

imkin replied:
Ah ok.
In Create, the default quota size is in d.options.quota.Size, where-as driver.quota.Size is initialized by the Create function to capture anything in "opts" param. In ReadOnly cases the opts will not have the "size" option since it does not go through the CreateReadWrite pathway.

cpuguy83 replied:
Can we actually check the `size` param instead of if the user passed any options?
I can see this causing a bug later.

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -343,20 +363,23 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
 		}
 	}()
 
+	driver := &Driver{}
```
All the changes from here on seem no longer needed. and comments are probably out dated as well.

amir73il commented:
@imkin better update the title and description of PR to overlay.size

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "30491-default-container-size" git@github.com:imkin/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354463224
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

imkin commented:
@GordonTheTurtle I am already doing that. Or can you point to what exactly am I missing?

imkin commented:
@amir73il I made the changes we discussed. Thanks

amir73il commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -345,14 +365,16 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
 			// Set container disk quota limit
 			if err := d.quotaCtl.SetQuota(dir, driver.options.quota); err != nil {
-				return err
+				return
```
??? please undo these removed lines and buggy return. Otherewise, LGTM

imkin replied:
Thanks for the catch. that was a bad miss.

imkin commented:
The failure of the windows test is not related to my change.

imkin commented:
@cpuguy83 I finished the changes as per the design review suggestions.  The failed test seems to be a flake or unrelated to my change.

cpuguy83 reviewed (changes requested):
Moving to code review
One nit.

stevvooe commented:
Should this be namespaced to make it clear that this is a project quota option? Another possible option is to make this option common to all implementations, as it really has nothing to do with overlay.

imkin commented:
@cpuguy83 made the change.
@stevvooe agree that the fix has nothing to do with overlay2. However I Think till all the graph drivers do no support this options naming it as overlay2.size is better. 
Project Quota is an implementation detail. I would doing any namespace around it.

imkin commented:
`######################################################################## 100.0%
08:35:47 curl: (22) The requested URL returned error: 400 Bad Request
08:35:47 Removing intermediate container 77f86e12d30d
08:35:47 The command '/bin/sh -c ./contrib/download-frozen-image-v2.sh /docker-frozen-images 	buildpack-deps:jessie@sha256:85b379ec16065e4fe4127eb1c5fb1bcc03c559bd36dbb2e22ff496de55925fa6 	busybox:latest@sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f 	debian:jessie@sha256:72f784399fd2719b4cb4e16ef8e369a39dc67f53d978cd3e2e7bf4e502c7b793 	hello-world:latest@sha256:c5515758d4c5e1e838e9cd307f6c6a0d620b5e07e6f927b07d05f6d12a1ac8d7' returned a non-zero code: 22
08:35:47 Build step 'Execute shell' marked build as failure
08:35:47 Set build name.
08:35:47 New build name is '#32977'
08:35:47 Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#32977', new value: '#32977'
08:35:47 [PostBuildScript] - Execution post build scripts.
08:35:47 [Docker-PRs] $ /bin/sh -xe /tmp/hudson6641650937514771287.sh
08:35:47 + docker stop docker-pr43374
08:35:48 Error response from daemon: No such container: docker-pr43374
08:35:48 + true
08:35:48 + id -u
08:35:48 + id -g
08:35:48 + docker run --rm -v /home/ubuntu/workspace/Docker-PRs:/workspace busybox chown -R 1000:1000 /workspace
08:35:48 [Docker-PRs] $ /bin/sh -xe /tmp/hudson8331392964947281436.sh
08:35:48 + + find bundles -name *.log -o -name *.prof -o -name integration.test
08:35:48 xargsfind: `bundles': No such file or directory
08:35:48  tar -czf bundles.tar.gz
08:35:48 tar: Cowardly refusing to create an empty archive
08:35:48 Try 'tar --help' or 'tar --usage' for more information.
08:35:48 + true
08:35:48 [Docker-PRs] $ /bin/sh -xe /tmp/hudson8034844993744393103.sh
08:35:48 + set +x
08:35:48 
08:35:48 
08:35:48 ---
08:35:48 Now starting POST-BUILD steps
08:35:48 ---
08:35:48 
08:35:49 Error response from daemon: No such container: docker-pr43374
08:35:49 Archiving artifacts
08:35:49 Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
08:35:50 [WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
08:35:50 Finished: FAILURE`

Do not think the failure in the experimental suite is due to the changes here.

thaJeztah commented:
@imkin that error is not related indeed; we're investigating it but haven't found yet what's causing it

imkin commented:
Since the failure of the experimental is not due to any changes here, Is this good for a another review?

thaJeztah commented:
ping @dmcgowan @cpuguy83 @stevvooe it's all 💚 now 😇

imkin commented:
@cpuguy83 I did make the changes as suggested.

thaJeztah commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -345,10 +369,12 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
 
 	if opts != nil && len(opts.StorageOpt) > 0 {
 		driver := &Driver{}
+
```
Can you remove this line? Looks unrelated to this change

imkin replied:
Done.

cpuguy83 commented on docs/reference/commandline/dockerd.md:
```diff
@@ -740,6 +740,19 @@ to add multiple lower directory support for OverlayFS. This option should
 
+##### `overlay2.size`
+
+Sets the default size of the container. It is supported only when the
```
`Sets the default size of the container` does not sound right.
Maybe `Sets the default max size for containers`

imkin replied:
Will put this in a different PR as the doc file has moved repos.

thaJeztah commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -345,10 +369,12 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
 			return err
 		}
 
+		// if the quota Size is set then attempt to set it.
```
This comment basically describes the comment directly in the `if` ("Set container disk quota limit"); I'd remove it because it's not really needed

imkin replied:
Done.

cpuguy83 commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -312,17 +323,30 @@ func (d *Driver) Cleanup() error {
+		if opts.StorageOpt == nil {
+			opts.StorageOpt = map[string]string{}
+		}
+		opts.StorageOpt["size"] = fmt.Sprintf("%d", d.options.quota.Size)
```
Can we use `strconv` for this? https://golang.org/pkg/strconv/

imkin replied:
Done.

thaJeztah commented on docs/reference/commandline/dockerd.md:
```diff
@@ -740,6 +740,19 @@ to add multiple lower directory support for OverlayFS. This option should
 only be used after verifying this support exists in the kernel. Applying
 this option on a kernel without this support will cause failures on mount.
 
+##### `overlay2.size`
```
Ohhh, the CLI reference was moved to the https://github.com/docker/cli repository (sorry! it was a long time in the making, and we finally were able to move them). Although `dockerd` itself is in this repository, this document was moved there as well; https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md

Could you open a separate pull-request in that repository, and remove the change here?

imkin replied:
Is there a timeline/plan when/if the dockerd will be moved?

thaJeztah replied:
`dockerd` itself will stay here (at least for the time being), only the client CLI is moved to the docker/cli repo

thaJeztah reviewed (changes requested):
Just spotted some nits (didn't do a "technical" review of the other parts)

imkin commented:
The failures do not seem to be connected to my changes.

```
21:42:51 curl: (22) The requested URL returned error: 400 Bad Request
21:42:51 Removing intermediate container 1b37e22497a7
21:42:51 The command '/bin/sh -c ./contrib/download-frozen-image-v2.sh /docker-frozen-images 	buildpack-deps:jessie@sha256:85b379ec16065e4fe4127eb1c5fb1bcc03c559bd36dbb2e22ff496de55925fa6 	busybox:latest@sha256:32f093055929dbc23dec4d03e09dfe971f5973a9ca5cf059cbfb644c206aa83f 	debian:jessie@sha256:72f784399fd2719b4cb4e16ef8e369a39dc67f53d978cd3e2e7bf4e502c7b793 	hello-world:latest@sha256:c5515758d4c5e1e838e9cd307f6c6a0d620b5e07e6f927b07d05f6d12a1ac8d7' returned a non-zero code: 22
21:42:51 Build step 'Execute shell' marked build as failure
21:42:51 Set build name.
21:42:51 New build name is '#32977'
21:42:51 Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#32977', new value: '#32977'
21:42:51 [PostBuildScript] - Execution post build scripts.
21:42:51 [Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson6029441601236704067.sh
21:42:51 + docker stop docker-pr-exp35037
21:42:51 Error response from daemon: No such container: docker-pr-exp35037
```

imkin commented:
@thaJeztah Please let me know if you are fine with the changes.
@cpuguy83 Made the requested changes.

cpuguy83 reviewed (approved):
LGTM

imkin commented:
@thaJeztah here is the docs PR https://github.com/docker/cli/pull/215

thaJeztah commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -312,17 +323,30 @@ func (d *Driver) Cleanup() error {
+		}
+	}
+
+	if _, ok := opts.StorageOpt["size"]; !ok {
```
Looks like we now _always_ pass `opts`, even if no option is set on `dockerd` or is passed with `docker run`; should the code here check if _either_ `opts.StorageOpt` is set _or_ `d.options.quota.Size` is set, and otherwise just keep `opts` as `nil`?

imkin replied:
When we reach CreateReadWrite we have lost the information whether opts.StorageOpt is defined or not. We only have d.options.quota.Size. Unintialized "d.options.quota.Size" is 0. I am not sure if it is semantically right to assume d.options.quota.Size with "0" as "Not-Initialized". The current "size" option for docker run allows you to define this as "0".

thaJeztah reviewed (changes requested):
left some questions

imkin commented:
@thaJeztah I could not comment on your second question so paraphrasing you here.
```
So if we don't want Create to be called with quota; perhaps we should add a non-exported create() that's both called by Create() and CreateReadWrite();

func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr error) {
	if opts != nil && len(opts.StorageOpt) != 0 && !projectQuotaSupported {
		return fmt.Errorf("--storage-opt is supported only for overlay over xfs with 'pquota' mount option")
	}

	return d.create(id, parent string, opts)
}
The check in Create() could even be;

if opts != nil && len(opts.StorageOpt) != 0 {
    // error...
}
Given that Create() then won't support any options
```

The check of "projectquotasupport" is not a concern of "Create()" but of CreateReadWrite() since only quota should only be set in the ReadWrite layer. Hence the check is moved to CreateReadWrite.

thaJeztah commented:
> The check of "projectquotasupport" is not a concern of "Create()" but of CreateReadWrite() since only quota should only be set in the ReadWrite layer. Hence the check is moved to CreateReadWrite.

The point is that `Create()` is an exported function, and, while true that it's currently not called with a quota; being an exported function means it _can_ be called in a different code path.

By splitting the Exported `Create()`, that situation would be covered

bklau commented:
@thaJeztah : Q: How is the default for container storage set for other drivers like dm , btrfs and zfs? Is this just a one-off for  Overlay2 driver only??

imkin commented:
@thaJeztah quota support does not make sense for read layers because they are anyways "readonly". Calling `Create` directly for setting quota may never be the right way. I could have a private `create` function that does what Create is doing today. As per your suggestion both Create and CreateReadWrite could call it. However the check `Create` could be to completely disallow setting quota directly.

bklau commented:
@imkin @rhvgoyal Earlier goyal did a test that shows that the size parameter also affects the "base image". On the client side, i thought that it just impacted the CoW RW layers.
Isn't this behaviour a discrepancy between what is set on by docker run --storage-opt size=xxx and dockerd --storage-opt overlay2.size=xxx ?? Pls clarify. I refer to below chat:

"@rhvgoyal just now did a docker pull of a image larger than what the overlay2.size parameter allowed.

bash-4.3# docker pull openjdk
Using default tag: latest
latest: Pulling from library/openjdk
cd0a524342ef: Extracting [==================================================>] 52.55 MB/52.55 MB
e39c3ffe4133: Download complete
85334a7c2001: Download complete
b46c5b79125e: Download complete
30d4cb7cc8bc: Download complete
4d273117faaf: Download complete
2b271bed4ca3: Download complete
30e4d8ac7a54: Download complete
45858d68b18d: Download complete
failed to register layer: Error processing tar file(exit status 1): open /usr/share/locale/nb/LC_MESSAGES/tar.mo: disk quota exceeded

So @rhvgoyal @amir73il it docker pull also honors the size parameter. The pull is successful when I increase the value of the "size" param."

imkin commented:
@bklau Initially the idea was to impact all layers. Then based on comments from @dmcgowan and @cpuguy83 we reduced the impact to be on ReadWrite Layer.(rootfs)

imkin commented:
@thaJeztah made the changes as per the discussion. Can you check if it satisfies your comments?

thaJeztah reviewed (approved):
LGTM thanks!

stevvooe commented:
@thaJeztah Wish we had taken the feedback in https://github.com/moby/moby/pull/32977#issuecomment-306284197 before merging this. The parameter `overlay2.size` is extremely unclear that it is a quota. What if soft and hard quotas are added? What if a different quota system is used to enforce that quota?

cpuguy83 commented:
@stevvooe This is not a new option, only able to specified at the daemon level in addition to per-container.

But sorry, your suggestion definitely got lost in the sea of comments. Review->RequestChanges may help here in the future.
Please open an issue if you think this should still be changed.

stevvooe commented:
`overlay2.quota.size` seems correct.

imkin commented:
@stevvooe I did respond to you but did not hear back from you. 

I am not sure why "quota" should be mentioned. Thats an implementation detail right?

stevvooe commented:
> I am not sure why "quota" should be mentioned. Thats an implementation detail right?

"Project Quota" is an implementation detail, yes, but the semantic meaning of `size` is that this is enforcing a quota. The fact that it is a quota is not an implementation detail.

The conversation is happening in #33903.
